File: gtci/14-islands-matrix-traversal/python3/1-number-of-islands.py
class Solution:
    def countIslands(self, matrix):
        WATER_DEF = 0
        ISLAND_DEF = 1

        N_ROWS = len(matrix)
        N_COLS = len(matrix[0])

        def dfs(row, col):
            stack = [(row, col)]

            while stack:
                [row, col] = stack.pop()

                # mark the cell as a water cell to indicate visited
                matrix[row][col] = WATER_DEF

                if row - 1 >= 0 and matrix[row - 1][col] == ISLAND_DEF:
                    stack.append((row - 1, col))

                if row + 1 < N_ROWS and matrix[row + 1][col] == ISLAND_DEF:
                    stack.append((row + 1, col))

                if col - 1 >= 0 and matrix[row][col - 1] == ISLAND_DEF:
                    stack.append((row, col - 1))

                if col + 1 < N_COLS and matrix[row][col + 1] == ISLAND_DEF:
                    stack.append((row, col + 1))

        totalIslands = 0
        for i in range(N_ROWS):
            for j in range(N_COLS):
                if matrix[i][j] == ISLAND_DEF:
                    totalIslands += 1
                    dfs(i, j)

        return totalIslands


if __name__ == "__main__":
    sol = Solution()

    assert (
        sol.countIslands(
            [
                [1, 1, 1, 0, 0],
                [0, 1, 0, 0, 1],
                [0, 0, 1, 1, 0],
                [0, 0, 1, 0, 0],
                [0, 0, 1, 0, 0],
            ]
        )
        == 3
    )

    assert (
        sol.countIslands(
            [
                [0, 1, 1, 1, 0],
                [0, 0, 0, 1, 1],
                [0, 1, 1, 1, 0],
                [0, 1, 1, 0, 0],
                [0, 0, 0, 0, 0],
            ]
        )
        == 1
    )

    print("All test cases passed.")

File: gtci/14-islands-matrix-traversal/cpp/1-number-of-islands.cpp
#include <cassert>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

class Solution
{
public:
  static constexpr int WATER_DEF = 0;
  static constexpr int LAND_DEF = 1;

  int countIslands(vector<vector<int>> &matrix)
  {
    const int N_ROWS = matrix.size();
    const int N_COLS = matrix[0].size();

    int totalIslands = 0;
    for (int i = 0; i < N_ROWS; i++) {
      for (int j = 0; j < N_COLS; j++) {
        if (matrix[i][j] == LAND_DEF) {
          totalIslands++;
          exploreIsland(i, j, matrix);
        }
      }
    }

    return totalIslands;
  }

private:
  void exploreIsland(const int startRow, const int startCol, vector<vector<int>>& matrix) {
    std::queue<std::pair<int, int>> queue;
    queue.push({startRow, startCol});

    while (!queue.empty()) {
      const auto [row, col] = queue.front();
      queue.pop();

      matrix[row][col] = WATER_DEF;

      const std::vector<std::pair<int, int>> directions {
        {-1, 0}, {1, 0}, {0, -1}, {0, 1}
      };
      
      for (const auto& [dr, dc] : directions) {
        const int nr = row + dr;
        const int nc= col + dc;

        if (nr >= 0 && nr < matrix.size() && nc >= 0 && nc < matrix[0].size() && matrix[nr][nc] == LAND_DEF)
          queue.push({nr, nc});
      }
    }
  }
};

int main() {
    Solution sol;

     vector<vector<int>> vec = vector<vector<int>>{
        {1, 1, 1, 0, 0},
        {0, 1, 0, 0, 1},
        {0, 0, 1, 1, 0},
        {0, 0, 1, 0, 0},
        {0, 0, 1, 0, 0}};
    assert(sol.countIslands(vec) == 3);

    vec = {
        {0, 1, 1, 1, 0},
        {0, 0, 0, 1, 1},
        {0, 1, 1, 1, 0},
        {0, 1, 1, 0, 0},
        {0, 0, 0, 0, 0}};
    assert(sol.countIslands(vec) == 1);

    cout << "All test cases passed." << endl;

    return 0;
}
